### PR TITLE
Feature/11 cob convert

### LINKED ISSUE #11: Center of brightness converter module
Develop module that inputs the Center of Brightness in pixels from an image and outputs a unit vector measurement

### PR DESCRIPTION
**Pull request by thibaudteil**
* **Tickets addressed:** lasp-bsk-011
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
Center of brightness conversion module. This module is a composable measurement model for the filters in flight software. 
It reads in the pixel values from image processing, and outputs the position of the target in frames of interest.

## Verification
Unit testing and CI server run.

## Documentation
Module was documented with rst file.

## Future work
None

### CODE REVIEW

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+/*! During an update, this module transforms pixel values for the center of brightness into a unit vector
+ * direction in several frames (inertial, Camera, and Body).
+ @return void
+ @param CurrentSimNanos The clock time at which the function was called (nanoseconds)
+ */
+void CobConverter::UpdateState(uint64_t CurrentSimNanos)
+{
+    std::string dirName;
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: :small_red_triangle_down: Unused variable?

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    
+public:
+    Message<OpNavUnitVecMsgPayload> opnavUnitVecOutMsg;
+    ReadFunctor<OpNavCOBMsgPayload> opnavCOBInMsg;
+    ReadFunctor<CameraConfigMsgPayload> cameraConfigInMsg;
+    ReadFunctor<NavAttMsgPayload> navAttInMsg;
+
+    uint64_t sensorTimeTag;
+    BSKLogger bskLogger;               
+
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.h]: :large_orange_diamond: Can remove this extra new line.

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.i [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+%include "architecture/msgPayloadDefCpp/OpNavUnitVecMsgPayload.h"
+%include "architecture/msgPayloadDefCpp/OpNavCOBMsgPayload.h"
+
+
+%pythoncode %{
+import sys
+protectAllClasses(sys.modules[__name__])
+%}
+
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.i]: :large_orange_diamond: Remove extra new line (leaving one at the end of the file). [If curious ](https://gist.github.com/OleksiyRudenko/d51388345ea55767b7672307fe35adf3)

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        eigenVector3d2CArray(rhat_BN_N, uVecMsgBuffer.rhat_BN_N);
+        eigenVector3d2CArray(rhat_BN_C, uVecMsgBuffer.rhat_BN_C);
+        eigenVector3d2CArray(rhat_BN_B, uVecMsgBuffer.rhat_BN_B);
+
+        uVecMsgBuffer.timeTag = (double) cobMsgBuffer.timeTag;
+        uVecMsgBuffer.valid = true;
+    }
+
+    this->opnavUnitVecOutMsg.write(&uVecMsgBuffer, this->moduleID, CurrentSimNanos);
+
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp:117]: :large_orange_diamond: Can remove this extra new line.

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        double X, Y;
+        double pX, pY;
+        /* compute sensorSize/focalLength = 2*tan(FOV/2) */
+        pX = 2.*tan(cameraSpecs.fieldOfView*cameraSpecs.resolution[0]/cameraSpecs.resolution[1]/2.0);
+        pY = 2.*tan(cameraSpecs.fieldOfView/2.0);
+        X = pX/cameraSpecs.resolution[0];
+        Y = pY/cameraSpecs.resolution[1];
+
+        /*! - Get the heading in the image plane */
+        Eigen::Matrix3d covar_N, covar_C, covar_B;
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: :large_orange_diamond: I would declare these closer to their usage site/assigning. But I may be missing something.

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        double CB[3][3], BN[3][3];
+        MRP2C(cameraSpecs.sigma_CB, CB);
+        dcm_CB = c2DArray2EigenMatrix3d(CB);
+        MRP2C(navAttBuffer.sigma_BN, BN);
+        dcm_BN = c2DArray2EigenMatrix3d(BN);
+
+        dcm_NC = dcm_BN.transpose() * dcm_CB.transpose();
+
+        /*! - Find pixel size using camera specs */
+        double X, Y;
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: :large_blue_circle: if you want, you could declare and assign these in one line.

File: src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+#  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+#  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+#  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+#  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+#  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+#  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+#  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+#
+
+from Basilisk.utilities import SimulationBaseClass, unitTestSupport, macros
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py]: :large_orange_diamond: Pycharm has a Code > Optimize Imports function which will order the imports. General to package local, alphabetically.

File: src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+@pytest.mark.parametrize("cameraResolution, sigma_CB, sigma_BN, centerOfBrightness, numberOfPixels",
+    [([512, 512], [1., 0.3, 0.1], [0.6, 1., 0.1], [152, 251], 75)
+    ,([128, 345], [0, 0, 0], [0, 0, 0], [251, 120], 100)
+    ,([742, 512], [-1., -0.3, -0.1], [-0.6, -1., -0.1], [0, 0], 0)
+    ,([2048, 2048], [-1., -0.3, -0.1], [-0.6, -1., -0.1], [1021, 1891], 1000)
+    ])
+
+def test_cob_converter(show_plots, cameraResolution, sigma_CB, sigma_BN, centerOfBrightness, numberOfPixels):
+    [testResults, testMessage] = cobConverterTestFunction(show_plots, cameraResolution, sigma_CB, sigma_BN, centerOfBrightness, numberOfPixels)
```
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py]: :large_orange_diamond: Can drop each parameter onto a new line.

```
[testResults, testMessage] = cobConverterTestFunction(show_plots, 
                                                      cameraResolution, 
                                                      sigma_CB, 
                                                      sigma_BN,                                                        
                                                      centerOfBrightness,                                                        
                                                      numberOfPixels)
```

**Review by patkenneally — APPROVED** [review]

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp (hunk 1 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: Fixed

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.h (hunk 2 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.h]: Fixed

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.i (hunk 3 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.i]: Fixed, thanks for the link

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp (hunk 4 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp:117]: fixed

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp (hunk 5 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp (hunk 6 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/cobConverter/cobConverter.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py (hunk 7 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py]: Done, we should run that on the whole project!

File: src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py (hunk 8 above)
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/cobConverter/_UnitTest/test_cobConverter.py]: Done

**Pull request merged**